apis/v0/jwt.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v0
+
+import (
+	"fmt"
+
+	"github.com/lestrrat-go/jwx/v2/jwt"
+)
+
+const (
+	// jwtJustificationsKey is the key in the JWT where justifications are stored.
+	// Ideally this would be "justifications", but the RFC and various online
+	// resources recommend key names be as short as possible to keep the JWTs
+	// small. Akamai recommends less than 8 characters and Okta recommends less
+	// than 6.
+	jwtJustificationsKey string = "justs"
+)
+
+// WithTypedJustifications is an option for parsing JWTs that will convert
+// decode the [Justification] claims into the correct Go structure. If this is
+// not supplied, the claims will be "any" and future type assertions will fail.
+func WithTypedJustifications() jwt.ParseOption {
+	return jwt.WithTypedClaim(jwtJustificationsKey, []*Justification{})
+}
+
+// GetJustifications retrieves a copy of the justifications on the token. If the
+// token does not have any justifications, it returns an empty slice of
+// justifications. Modifying the slice does not modify the underlying token -
+// you must call SetJustifications to update the data on the token.
+func GetJustifications(t jwt.Token) ([]*Justification, error) {
+	if t == nil {
+		return nil, fmt.Errorf("token cannot be nil")
+	}
+
+	raw, ok := t.Get(jwtJustificationsKey)
+	if !ok {
+		return []*Justification{}, nil
+	}
+
+	typ, ok := raw.([]*Justification)
+	if !ok {
+		return nil, fmt.Errorf("found justifications, but was %T (expected %T)",
+			raw, []*Justification{})
+	}
+
+	// Make a copy of the slice so we don't modify the underlying data structure.
```

Comment:
This is actually a fun bug in the previous implementation, since modifying the returned slice would actually modify the token. The performance overhead of making these copies seems work the security trade-off risks of setting invalid justifications.